### PR TITLE
LibWeb: Selection::getRangeAt handle focus and anchor

### PR DESCRIPTION
**Pull request by shlyakpavel**
Copies https://github.com/LadybirdBrowser/ladybird/pull/1767
I rebased on master and rebaselined a failing test 
P. S. Reverted the test
CC @kalenikaliaksandr as the one actually writing the final code and @gmta as the one who knows the most about Selection API

### CODE REVIEW

File: Libraries/LibWeb/Selection/Selection.cpp [hunk 1]
```diff
@@ -134,8 +134,11 @@ String Selection::direction() const
 // https://w3c.github.io/selection-api/#dom-selection-getrangeat
 WebIDL::ExceptionOr<GC::Ptr<DOM::Range>> Selection::get_range_at(unsigned index)
 {
-    // The method must throw an IndexSizeError exception if index is not 0, or if this is empty.
-    if (index != 0 || is_empty())
+    // The method must throw an IndexSizeError exception if index is not 0, or if this is empty or either focus or anchor is not in the document tree.
+    auto is_focus_in_document_tree = focus_node() && focus_node()->in_a_document_tree();
+    auto is_anchor_in_document_tree = anchor_node() && anchor_node()->in_a_document_tree();
```
Comment (gmta) [inline Libraries/LibWeb/Selection/Selection.cpp:139]: The spec states "in _the_ document tree", not "in _a_ document tree". I think this is trying to confirm that both the anchor and focus are in the same document as the selection itself.

Comment (shlyakpavel) [reply Libraries/LibWeb/Selection/Selection.cpp]: I guess I resolved that

File: Libraries/LibWeb/Selection/Selection.cpp [hunk 2]
```diff
@@ -134,8 +135,15 @@ String Selection::direction() const
...
 {
-    // The method must throw an IndexSizeError exception if index is not 0, or if this is empty.
-    if (index != 0 || is_empty())
+    GC::Ptr<DOM::Document> selection_document = document();
+    GC::Ptr<DOM::Node> focus = focus_node();
+    GC::Ptr<DOM::Node> anchor = anchor_node();
+
+    // The method must throw an IndexSizeError exception if index is not 0, or if this is empty or either focus or anchor is not in the document tree.
+    auto is_focus_in_document_tree = focus && &focus->document() == selection_document;
+    auto is_anchor_in_document_tree = anchor && &anchor->document() == selection_document;
```
Comment (gmta) [inline Libraries/LibWeb/Selection/Selection.cpp:144]: The `document()` on line 138 gives you a `GC::Ref`, not a `GC::Ptr`, and you can directly compare these values here.

File: Libraries/LibWeb/Selection/Selection.cpp [hunk 3]
```diff
@@ -8,6 +8,7 @@
 #include <LibWeb/Bindings/Intrinsics.h>
```
Comment (gmta) [inline Libraries/LibWeb/Selection/Selection.cpp:8]: Commit message:
> LibWeb: Selection::getRangeAt handle focus and anchor

Make that a sentence, e.g.
> LibWeb: Update Selection::getRangeAt() spec to handle focus and anchor

This is also not a complete sentence and lacks punctuation:
> Following a spec change, need to check for focus and anchor in range

Comment (shlyakpavel) [reply Libraries/LibWeb/Selection/Selection.cpp:8]: Yeah, that's a leftover from the previous committer :)
Fixed

Comment (gmta) [reply Libraries/LibWeb/Selection/Selection.cpp:8]: Oh that's right, in that case you'd mention both authors with `Co-authored-by` tags in your commit message :)

Comment (shlyakpavel) [reply Libraries/LibWeb/Selection/Selection.cpp:8]: There's nothing left from the previous PR in this PR
Not a line of code :)

Comment (shlyakpavel) [reply Libraries/LibWeb/Selection/Selection.cpp:8]: It includes a line proposed by @kalenikaliaksandr at https://github.com/LadybirdBrowser/ladybird/pull/1767#discussion_r1798470277
I'm not sure if I should mention them as coauthor

Comment (shlyakpavel) [reply Libraries/LibWeb/Selection/Selection.cpp:8]: Well who cares anyway let's list them all and rock

**Pull request merged**